Failure in TestExtractStuffAndWhy with Twisted 17.9

Hello,

Following bug [#876988 in Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=876988), when I try to run tests with Twisted 17.9 installed, I get the following error:

```
================================================= FAILURES =================================================
_______________________________ TestExtractStuffAndWhy.test_handlesFailures ________________________________

self = <tests.test_twisted.TestExtractStuffAndWhy object at 0x7f2a8a4917d0>

    def test_handlesFailures(self):
        """
            Extracts failures and events.
            """
        f = Failure(ValueError())
>       assert (
            (f, "foo", {}) ==
            _extractStuffAndWhy({"_why": "foo",
                                 "_stuff": f})
        )
E       AssertionError: assert (<twisted.pyt... >, 'foo', {}) == (<twisted.pyth... >, 'foo', {})
E         At index 0 diff: <twisted.python.failure.Failure exceptions.ValueError: > != <twisted.python.failure.Failure exceptions.ValueError: >
E         Full diff:
E         (<twisted.python.failure.Failure exceptions.ValueError: >, 'foo', {})

tests/test_twisted.py:117: AssertionError
============================= 1 failed, 229 passed, 10 skipped in 0.50 seconds =============================
```

Putting some print arounds, the problem is that `isinstance(_stuff, BaseException)` is true, so we create a new encapsulate `_stuff` twice. I'll propose a fix shortly.

--- a/.travis.yml
+++ b/.travis.yml
@@ -32,6 +32,8 @@ matrix:
       env: TOXENV=py27-colorama
     - python: "3.6"
       env: TOXENV=py36-colorama
+    - python: "3.6"
+      env: TOXENV=py36-oldtwisted
 
 
     # Meta

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -26,7 +26,8 @@ Changes:
 
 - Empty strings are valid events now.
   `#110 <https://github.com/hynek/structlog/issues/110>`_
-
+- Do not encapsulate Twisted failures twice with newer versions of Twisted.
+  `#144 <https://github.com/hynek/structlog/issues/144>`_
 
 ----
 

--- a/src/structlog/twisted.py
+++ b/src/structlog/twisted.py
@@ -107,7 +107,7 @@ def _extractStuffAndWhy(eventDict):
         _stuff = Failure()
     # Either we used the error ourselves or the user supplied one for
     # formatting.  Avoid log.err() to dump another traceback into the log.
-    if isinstance(_stuff, BaseException):
+    if isinstance(_stuff, BaseException) and not isinstance(_stuff, Failure):
         _stuff = Failure(_stuff)
     if PY2:
         sys.exc_clear()

--- a/tox.ini
+++ b/tox.ini
@@ -1,18 +1,18 @@
 [tox]
-envlist = {py27,py34,py35,py36,pypy}-{threads,greenlets},{py27,py36}-colorama,flake8,docs,readme,manifest,coverage-report
+envlist = {py27,py34,py35,py36,pypy}-{threads,greenlets},{py27,py36}-{colorama,oldtwisted},flake8,docs,readme,manifest,coverage-report
 
 
 [testenv]
 deps =
     -rdev-requirements.txt
     greenlets: greenlet
-    twisted
+    threads,greenlets,colorama: twisted
+    oldtwisted: twisted < 17
     colorama: colorama
     py36: python-rapidjson
 setenv =
     PYTHONHASHSEED = 0
 commands = coverage run --parallel -m pytest {posargs}
-
 
 [testenv:flake8]
 basepython = python3.6

